src/mbedtls_socket.cc

```diff
@@ -261,7 +275,9 @@ int MbedtlsSocket::Accept(int sockfd, sockaddr* addr, socklen_t* addrlen, int fl
       throw std::runtime_error("socket unavailable");
 
     re = mbedtls_ssl_handshake(&ctx.ssl);
-    if (re == MBEDTLS_ERR_SSL_WANT_READ || re == MBEDTLS_ERR_SSL_WANT_WRITE) {
+    if (re == MBEDTLS_ERR_SSL_FATAL_ALERT_MESSAGE) {  // close connection before throwing
```

Comment:
Right, we probably want to Close() every connection for which we return ECONNABORTED. Do we want to close the connection on CheckResult then? This means that CheckResult needs the context (e.g. the fd). Since this is only relevant for Accept(), I think it is ok to just check the return once in Accept() and then again in CheckResult(). With this approach we also can handle a cert parsing error in Connect() different then in Accept(). This is probably needed at some point anyway.

Reply:
> check the return once in Accept() and then again in CheckResult()

I agree.

> we probably want to Close() every connection for which we return ECONNABORTED

Or maybe we want to close (and remove from contexts_) whenever we throw?

Reply:
> Or maybe we want to close (and remove from contexts_) whenever we throw?
Do you mean in every case (e.g. also write/reads?). My hope is that the program using ttls checks for errors there. I generally don't want to close, where I don't have to, to mimic the syscalls as close as possible. My hope is, that if the program recovers from the error, it closes the connection/fd itself.
I think Accept() is the only exception because we accept a new connection that only we know about and can notify the program that no new connection was established (e.g. die to an mbedtls error).

Reply:
> for which we return ECONNABORTED

My point here was that there are other MBEDTLS errors where we don't set ECONNABORTED, but still throw (in accept).

Reply:
Also, we should generally refactor this to throw `system_error`s with errnos and let the dispatcher catch and set errno. And let it set some default errno for other exceptions that we didn't expect.

Reply:
This is true. The program can't close a fd which we never returned of course. Sorry for the overly eager fix. 36d2f2d  Should contain an adequate solution. Will cleanup before merge of course.